app/src/main/java/org/mozilla/rocket/banner/BannerAdapter.java

```
@@ -29,27 +33,39 @@ public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
     }
 
     public BannerAdapter(@NonNull String[] configs, OnClickListener onClickListener) throws JSONException {
-        this.DAOs = new BannerDAO[configs.length];
+        this.DAOs = new ArrayList<>();
         this.onClickListener = onClickListener;
-        for (int i = 0; i < configs.length ; i++) {
-            JSONObject jsonObject = new JSONObject(configs[i]);
+        for (String config : configs) {
+            JSONObject jsonObject = new JSONObject(config);
             BannerDAO thisDAO = new BannerDAO();
             thisDAO.type = jsonObject.getString(BannerDAO.TYPE_KEY);
             thisDAO.values = jsonObject.getJSONArray(BannerDAO.VALUES_KEY);
-            this.DAOs[i] = thisDAO;
+            // We always write full item into cache, but discard a page if it is not know in this
+            // apk version. With this once the user upgrades, the cache is still usable.
+            // We still inflate an item with SingleButtonViewHolder if it for unknown reason managed
+            // to pass this check. (See: onCreateViewHolder)
+            if (getItemViewType(thisDAO.type) == UNKNOWN_VIEW_TYPE) {
+                continue;
```

Should we throw an exception on nightly?